### ISSUE TITLE
Unable to get Horovod to run successfully on multiple-GPU

### ISSUE THREAD

**Issue by nischalhp** [uber/horovod#75]
I am currently running InceptionV3 model on a Tesla K80 GPU. And here is the code snippet that works perfectly fine:
        
            file_list = build_data(sys.argv[1])
            train_size = math.ceil(0.8 * len(file_list))

            train_file_list = file_list[:train_size]
            test_file_list=file_list[train_size:]

            nbatches_train, mod = divmod(len(train_file_list), BATCH_SIZE)
            nbatches_valid, mod = divmod(len(test_file_list), BATCH_SIZE)

            # Initialise session
            os.environ["CUDA_VISIBLE_DEVICES"] = "0"
            sess = tf.Session()
            K.set_session(sess)
            K.set_learning_phase(1)

            model = create_inception_model(num_classes=4, input_shape=(HEIGHT, WIDTH, CHANNELS),
                                           fully_trainable=True)

            cb = []

            tbCallBack = TensorBoard(log_dir='./Graph', histogram_freq=0,
                                     write_graph=True, write_images=True)
            cb.append(tbCallBack)

            train_generator = generator_from_data_path(train_file_list, batch_size=BATCH_SIZE, width=WIDTH, height=HEIGHT)
            test_generator = generator_from_data_path(test_file_list, batch_size=BATCH_SIZE, width=WIDTH, height=HEIGHT)

            model.fit_generator(train_generator, epochs=EPOCHS,
                                steps_per_epoch=nbatches_train,callbacks=cb,
                                validation_data=test_generator,
                                validation_steps=nbatches_valid)

            serialize_mode(model, name='inception')`

As always, the training takes a longer time and hence I wanted to shift to Horovod to run it on multiple GPU's.
Here is the code that I added to the above code in trying to make that happen:

Here is the addition to the model generation code:

    opt = keras.optimizers.Adadelta()
    opt = hvd.DistributedOptimizer(opt)

    model.compile(loss='sparse_categorical_crossentropy', optimizer=opt, metrics=['accuracy'])
    return model

And here is the code where we are initializing Horovod to run the model:

            hvd.init()
            config = tf.ConfigProto()
            config.gpu_options.allow_growth = True
            config.gpu_options.visible_device_list = str(hvd.local_rank())
            K.set_session(tf.Session(config=config))

            file_list = build_data(sys.argv[1])
            train_size = math.ceil(0.8 * len(file_list))

            train_file_list = file_list[:train_size]
            test_file_list=file_list[train_size:]

            nbatches_train, mod = divmod(len(train_file_list), BATCH_SIZE)
            nbatches_valid, mod = divmod(len(test_file_list), BATCH_SIZE)

            model = create_inception_model(num_classes=4, input_shape=(HEIGHT, WIDTH, CHANNELS),
                                           fully_trainable=True)

            cb = []

            tbCallBack = TensorBoard(log_dir='./Graph', histogram_freq=0,
                                     write_graph=True, write_images=True)
            if hvd.rank()  == 0:
                modelCheckPointCallBack = keras.callbacks.ModelCheckpoint('./checkpoint-{epoch}.h5')
                cb.append(modelCheckPointCallBack)

            cb.append(tbCallBack)

            train_generator = generator_from_data_path(train_file_list, batch_size=BATCH_SIZE, width=WIDTH, height=HEIGHT)
            test_generator = generator_from_data_path(test_file_list, batch_size=BATCH_SIZE, width=WIDTH, height=HEIGHT)

            model.fit_generator(train_generator, epochs=EPOCHS,
                                steps_per_epoch=nbatches_train,callbacks=cb,
                                validation_data=test_generator,
                                validation_steps=nbatches_valid)

            serialize_mode(model, name='inception')

And this is what happens when I run Horovod with 4 gpus:
```

--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 4 slots
that were requested by the application:
  python

Either request fewer slots for your application, or make more slots available
for use.
--------------------------------------------------------------------------
```
But when I run watch nvidia-smi I have 4 GPU's :

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.90                 Driver Version: 384.90                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                  Off |
| N/A   31C    P8    28W / 149W |     16MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           Off  | 00000000:00:05.0 Off |                  Off |
| N/A   31C    P8    28W / 149W |      1MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           Off  | 00000000:00:06.0 Off |                  Off |
| N/A   33C    P8    29W / 149W |      1MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           Off  | 00000000:00:07.0 Off |                  Off |
| N/A   30C    P8    27W / 149W |      1MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1956      G   /usr/lib/xorg/Xorg                            15MiB |
+-----------------------------------------------------------------------------+

```

When I run the same program with 1 gpu, my nvidia-smi command looks like this:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.90                 Driver Version: 384.90                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                  Off |
| N/A   42C    P0   140W / 149W |  11690MiB / 12205MiB |     82%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           Off  | 00000000:00:05.0 Off |                  Off |
| N/A   36C    P0    56W / 149W |  11598MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           Off  | 00000000:00:06.0 Off |                  Off |
| N/A   38C    P0    69W / 149W |  11598MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           Off  | 00000000:00:07.0 Off |                  Off |
| N/A   35C    P0    55W / 149W |  11598MiB / 12205MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1956      G   /usr/lib/xorg/Xorg                            15MiB |
|    0      3447      C   python                                     11661MiB |
|    1      3447      C   python                                     11585MiB |
|    2      3447      C   python                                     11585MiB |
|    3      3447      C   python                                     11585MiB |
+-----------------------------------------------------------------------------+
```

It creates 4 python processes to be used by GPU's but it does not run it in a distributed fashion.

How do I get this to run on a single machine with Multiple GPU's.

**Comment by alsrgv** [comment]
Can you try adding `—bind-to none` to your `mpirun` command?

**Comment by nischalhp** [comment]
@alsrgv I did, no change to the outcome.

**Comment by alsrgv** [comment]
@nischalhp, gotcha, here are couple more things to try:

1. Can you add `-oversubscribe` argument to `mpirun`. You can keep `-bind-to` as it still may be helpful.

2. Keras 2.0.9 has a bug (https://github.com/fchollet/keras/issues/8353) which causes it to grab all available GPUs. If you’re using Keras 2.0.9, can you downgrade to 2.0.8?

**Comment by nischalhp** [comment]
Adding `-oversubscribe` with the downgrade of the library worked. 
Thank you @alsrgv :) 
Will close the issue now.

**Comment by adiri** [comment]
@nischalhp which library did you downgrade? I have the same situation here.

**Comment by abditag2** [comment]
Keras 2.0.9 to  to 2.0.8
